nncf/pruning/model_analysis.py

```diff
@@ -0,0 +1,243 @@
+            cls = StopMaskForwardOps
+        return cls
+
+    def propagate_can_prune_attr_up(self):
+        """
+        Propagating can_prune attribute in reversed topological order.
+        This attribute depends on accept_pruned_input and can_prune attributes of output nodes.
+        Node can_prune is True if all outputs accept_pruned_input is True and all outputs
+        (except convs because conv can be pruned by input and output independently).
+        """
+        reversed_sorted_nodes = reversed([self.nx_graph.nodes[name] for name in nx.topological_sort(self.nx_graph)])
+        for nx_node in reversed_sorted_nodes:
+            nncf_node = self.graph._nx_node_to_nncf_node(nx_node)
+
+            # Check all output nodes accept_pruned_input attribute
+            out_nodes = self.graph.get_next_nodes(nncf_node)
+            outputs_accept_pruned_input = all(self.accept_pruned_input[node.node_id] for node in out_nodes)
+
+            # Check all output nodes can_prune attribute
+            outputs_will_be_pruned = all([self.can_prune[node.node_id]
+                                          for node in out_nodes if self.node_propagate_can_prune_attr(node)])
+            self.can_prune[nncf_node.node_id] = outputs_accept_pruned_input and outputs_will_be_pruned
+
+    def propagate_can_prune_attr_down(self):
+        """
+        Propagating can_prune attribute down to fix all branching cases with one pruned and one not pruned
+        branches.
+        """
+        sorted_nodes = [self.nx_graph.nodes[name] for name in nx.topological_sort(self.nx_graph)]
+        for nx_node in sorted_nodes:
+            nncf_node = self.graph._nx_node_to_nncf_node(nx_node)
+            # Propagate attribute only in not conv case
+            if self.node_propagate_can_prune_attr(nncf_node):
+                in_nodes = self.graph.get_previous_nodes(nncf_node)
+                can_prune = all([self.can_prune[node.node_id] for node in in_nodes])
+                can_prune_any = any([self.can_prune[node.node_id] for node in in_nodes])
+
+                if (not self.node_accept_different_inputs(nncf_node) and not can_prune) or \
+                        (self.node_accept_different_inputs(nncf_node) and not can_prune_any):
+                    self.can_prune[nncf_node.node_id] = can_prune
+
+    def set_accept_pruned_input_attr(self):
+        for nncf_node in self.graph.get_all_nodes():
+            node_module = self.model.get_module_by_scope(nncf_node.op_exec_context.scope_in_model)
+            node_type = nncf_node.op_exec_context.operator_name
+            cls = self.get_class_by_type_name(node_type)()
+            self.accept_pruned_input[nncf_node.node_id] = cls.accept_pruned_input(self.model, self.graph, node_module)
+
+    def analyse_model_before_pruning(self):
+        self.set_accept_pruned_input_attr()
+        self.propagate_can_prune_attr_up()
+        self.propagate_can_prune_attr_down()
+        return self.can_prune
```

Comment:
I assume that the final groups to be pruned are derived as a result of the interaction between the `Analyzer`, `Clusterization`, `NodesCluster` entities. There is quite a lot of code and logic here. I suggest that you unit-test the interaction, so that future maintainers could have some kind of a reference towards how this algorithm should work, and to avoid the situations in the field where the exportable pruning rate is less than the in-Pytorch pruning rate.

Reply:
I added unit-tests for all of the structures and functions that analyze the model and interaction of them when it is applicable. Please, check them in the `test_model_pruning_analysis.py` file.